src/Swarm/Game/State/Robot.hs

```diff
@@ -378,17 +381,66 @@ removeRobotFromLocationMap (Cosmic oldSubworld oldPlanar) rid =
   tidyDelete robID =
     surfaceEmpty M.null . M.update (deleteOne robID) oldPlanar
 
+setRobotInfo :: RID -> [Robot] -> Robots -> Robots
+setRobotInfo baseID robotList rState =
+  (setRobotList robotList rState) {_focusedRobotID = baseID}
+    & viewCenterRule .~ VCRobot baseID
+
 setRobotList :: [Robot] -> Robots -> Robots
 setRobotList robotList rState =
   rState
     & robotMap .~ IM.fromList (map (view robotID &&& id) robotList)
     & robotsByLocation .~ M.map (groupRobotsByPlanarLocation . NE.toList) (groupRobotsBySubworld robotList)
     & internalActiveRobots .~ setOf (traverse . robotID) robotList
+    & robotNaming . gensym .~ initGensym
  where
+  initGensym = length robotList - 1
+
   groupRobotsBySubworld =
     binTuples . map (view (robotLocation . subworld) &&& id)
 
   groupRobotsByPlanarLocation rs =
     M.fromListWith
       IS.union
       (map (view (robotLocation . planar) &&& (IS.singleton . view robotID)) rs)
+
+-- | Modify the 'viewCenter' by applying an arbitrary function to the
+--   current value.  Note that this also modifies the 'viewCenterRule'
+--   to match.  After calling this function the 'viewCenterRule' will
+--   specify a particular location, not a robot.
+modifyViewCenter :: (Cosmic Location -> Cosmic Location) -> Robots -> Robots
+modifyViewCenter update rInfo =
+  rInfo
+    & case rInfo ^. viewCenterRule of
+      VCLocation l -> viewCenterRule .~ VCLocation (update l)
+      VCRobot _ -> viewCenterRule .~ VCLocation (update (rInfo ^. viewCenter))
+
+-- | "Unfocus" by modifying the view center rule to look at the
+--   current location instead of a specific robot, and also set the
+--   focused robot ID to an invalid value.  In classic mode this
+--   causes the map view to become nothing but static.
+unfocus :: Robots -> Robots
+unfocus = (\ri -> ri {_focusedRobotID = -1000}) . modifyViewCenter id
+
+-- | Recalculate the view center (and cache the result in the
+--   'viewCenter' field) based on the current 'viewCenterRule'.  If
+--   the 'viewCenterRule' specifies a robot which does not exist,
+--   simply leave the current 'viewCenter' as it is. Set 'needsRedraw'
```

Comment:
This sentence about setting `needsRedraw` should be removed, right?  That's now handled in `recalcViewCenterAndRedraw`.